pkg/update/update.go

```diff
@@ -0,0 +1,158 @@
+package update
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/glasskube/glasskube/api/v1alpha1"
+	"github.com/glasskube/glasskube/internal/repo"
+	"github.com/glasskube/glasskube/pkg/client"
+	"github.com/glasskube/glasskube/pkg/condition"
+	"github.com/glasskube/glasskube/pkg/statuswriter"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type UpdateResult string
+
+const (
+	UpdateResultUnchanged UpdateResult = "unchanged"
+	UpdateResultUpdated   UpdateResult = "updated"
+)
+
+type UpdateTransaction struct {
+	Items []updateTransactionItem
+}
+
+func (tx UpdateTransaction) IsEmpty() bool {
+	for _, item := range tx.Items {
+		if item.Version != "" {
+			return false
+		}
+	}
+	return true
+}
+
+type updateTransactionItem struct {
+	Package v1alpha1.Package
+	Version string
+}
+
+func (txi updateTransactionItem) UpdateRequired() bool {
+	return txi.Version != ""
+}
+
+type updater struct {
+	client *client.PackageV1Alpha1Client
+	status statuswriter.StatusWriter
+}
+
+func NewUpdater(client *client.PackageV1Alpha1Client) *updater {
+	return &updater{
+		client: client,
+	}
+}
+
+func (c *updater) WithStatusWriter(writer statuswriter.StatusWriter) *updater {
+	c.status = writer
+	return c
+}
+
+func (c *updater) Prepare(ctx context.Context, packageNames []string) (*UpdateTransaction, error) {
+	c.status.Start()
+	defer c.status.Stop()
+	c.status.SetStatus("Collecting installed packages")
+	var packagesToUpdate []v1alpha1.Package
+	if len(packageNames) > 0 {
+		// Fetch all requested packages individually.
+		// This way, we can fail early if a requested package is not installed.
+		for _, name := range packageNames {
+			var pkg v1alpha1.Package
+			if err := c.client.Packages().Get(ctx, name, &pkg); err != nil {
+				return nil, fmt.Errorf("failed to get package %v: %v", name, err)
+			}
+			packagesToUpdate = append(packagesToUpdate, pkg)
+		}
+	} else {
+		var packageList v1alpha1.PackageList
+		if err := c.client.Packages().GetAll(ctx, &packageList); err != nil {
+			return nil, fmt.Errorf("failed to get list of installed packages: %v", err)
+		}
+		packagesToUpdate = packageList.Items
+	}
+
+	c.status.SetStatus("Updating package index")
+	var index repo.PackageRepoIndex
+	if err := repo.FetchPackageRepoIndex("", &index); err != nil {
+		return nil, fmt.Errorf("failed to fetch index: %v", err)
+	}
+
+	var tx UpdateTransaction
+outer:
+	for _, pkg := range packagesToUpdate {
+		for _, indexItem := range index.Packages {
```

Comment:
Since we only need the `LatestVersion`, I thought it would be more efficient to download the index file. As you say, it's only one request and the size per package is much smaller. Event with some 100-1000 packages, I don't think this would be a huge problem.

Once we reach an index size where this approach is no longer reasonable we can still change the updater but until then I think fetching the index should be fine.